include/ginkgo/core/matrix/csr_lookup.hpp

```diff
@@ -0,0 +1,186 @@
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_PUBLIC_CORE_MATRIX_CSR_LOOKUP_HPP_
+#define GKO_PUBLIC_CORE_MATRIX_CSR_LOOKUP_HPP_
+
+
+#include <type_traits>
+
+
+#include <ginkgo/core/base/intrinsics.hpp>
+#include <ginkgo/core/base/types.hpp>
+#include <ginkgo/core/matrix/csr.hpp>
+
+
+namespace gko {
+namespace matrix {
+
+
+/**
+ * Type describing which kind of lookup structure is used to find entries in a
+ * single row of a Csr matrix.
+ */
+enum class sparsity_type : int {
+    /**
+     * The row is dense, i.e. it contains all entries in
+     * `[min_col, min_col + storage_size)`.
+     * This means that the relative output index is `col - min_col`.
+     */
+    full = 1,
+    /**
+     * The row is sufficiently dense that its sparsity pattern can be stored in
+     * a dense bitmap consisting of `storage_size` blocks of size `block_size`,
+     * with a total of `storage_size` blocks. Each block stores its sparsity
+     * pattern and the number of columns before. This means that the relative
+     * output index can be computed as
+     * ```
+     * auto block = (col - min_col) / block_size;
+     * auto local_col = (col - min_col) % block_size;
+     * auto prefix_mask = (block_type{1} << local_col) - 1;
+     * auto output_idx = base[block] + popcount(bitmap[block] & prefix_mask);
+     * ```
+     */
+    bitmap = 2,
+    /**
+     * The row is sparse, so it is best represented using a hashtable.
+     * The hashtable has size `storage_size` and stores the relative output
+     * index directly, i.e.
+     * ```
+     * auto hash_key = col - min_col;
+     * auto hash_bucket = hash(hash_key);
+     * while (col_idxs[hashtable[hash_bucket]] != col) {
+     *     hash_bucket = (hash_bucket + 1) % storage_size; // linear probing
+     * }
+     * auto output_idx = hashtable[hash_bucket];
+     * ```
+     */
+    hash = 4,
+};
+
+
+inline sparsity_type operator|(sparsity_type a, sparsity_type b)
+{
+    return static_cast<sparsity_type>(static_cast<int>(a) |
+                                      static_cast<int>(b));
+}
+
+
+template <typename IndexType>
+class csr_sparsity_lookup {
```

Comment:
What is the purpose of this class? It is not used anywhere, and it has only private members.

Reply:
Should the rest of this file be part of the public interface? Or would it be better to place this file under `core`?